{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "05fa21de",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mat_dataset import *\n",
    "import numpy as np\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from math import pow    \n",
    "# import fastai\n",
    "\n",
    "# ----------------------------------------------------------------------------------\n",
    "# Functions to handle the unique attributes of the consensus1.mat file in Gong Lab\n",
    "# @author: davidebear\n",
    "# \n",
    "# Note: Refer to https://github.com/emiliosalazar/matToPython/blob/main/MatFileMethods.py for a more encompassing MatLabLoader\n",
    "# ----------------------------------------------------------------------------------\n",
    "\n",
    "# Needed imports\n",
    "import numpy as np\n",
    "import h5py\n",
    "\n",
    "def load_matlab_file(mat_file, variable_name): # Note the double return statement\n",
    "    \"\"\"_summary_\n",
    "    //UNIQUE: Only one variable for this .mat file\n",
    "    \n",
    "    Args:\n",
    "        mat_file (string): Name of .mat file to intake (path must be given if not in current dir)\n",
    "        variable_name (string): which variable from the mat file. Only one here because of the .mat file structure.\n",
    "\n",
    "    Returns:\n",
    "        h5py_object, data: Overall h5py_object and the specific variable as an nparray\n",
    "    \"\"\"\n",
    "    h5py_object = h5py.File(mat_file, 'r')\n",
    "    data = h5py_object.get(variable_name)\n",
    "    data = np.array(data)\n",
    "    return h5py_object, data\n",
    "\n",
    "class DNA_SeqBlocks():\n",
    "    \"\"\"_summary_\n",
    "    Loads the mat file with DNA sequences for easy iteration through all stored DNA sequence lists of varying coverage\n",
    "    \"\"\"\n",
    "    def __init__(self, h5py_object, data):\n",
    "        \"\"\"_summary_\n",
    "        Args:\n",
    "            h5py_object (h5py_object): first return variable of load_matlab_file\n",
    "            data (np.array): second return variable of load_matlab_file\n",
    "        \"\"\"\n",
    "        self.h5py_object = h5py_object\n",
    "        self.data = np.transpose(data)\n",
    "        self.size = len(data) \n",
    "    def get_seqblock(self, number): # //TODO: make sure the default args work\n",
    "        x = self.h5py_object[self.data[0, number-1]][:, :] # //UNIQUE: 0th column because .mat file, align3 variable only has one row\n",
    "        return np.transpose(x)\n",
    "    \n",
    "# CURRENT ISSUE:  I'm getting an int array instead of a char array. Use pd dataframe? How do I force a char?\n",
    "def seqblock_parser(seqblock):\n",
    "    seqblock_parsed = CleanSeqBlock()\n",
    "    total_columns = len(np.transpose(seqblock)) # is there a more efficient way?\n",
    "    total_rows = len(seqblock)\n",
    "    number_of_reads = int((total_rows - 4)/2) # last three rows have intreptations, first row is non-mutated target sequence, and N quality scores for N reads\n",
    "    \n",
    "    seqblock_parsed.reads_count = number_of_reads;\n",
    "    seqblock_parsed.seq_len = total_columns;\n",
    "    \n",
    "    # print(total_columns)\n",
    "    # print(total_rows)\n",
    "    \n",
    "    for i in range(0,total_rows-1):\n",
    "        # print(f\" The sequenceblock input {seqblock}\") #debug\n",
    "        # print(f\" The m x n size of the seqblock {seqblock.shape}\") #debug\n",
    "        # print(f\" The ith row of the seqblock {current_seq}\") #debug\n",
    "        # print(f\" The m x n size of the sequence {current_seq.shape}\") #debug\n",
    "        new_format = np.chararray(total_columns)\n",
    "        new_format[:] = 'q' #debug\n",
    "        \n",
    "        # print(f\" The 0th entry of the initialized char array {new_format[0, 0]}\") #debug\n",
    "        # print(f\" Its row size {len(new_format[0])}\")\n",
    "        for j in range(total_columns):\n",
    "            # print(j)\n",
    "            new_format[j] = chr(seqblock[i,j])\n",
    "            \n",
    "            \n",
    "        if (i == 0):\n",
    "            seqblock_parsed.target = new_format\n",
    "        elif (i == total_rows - 1): # last row is mutations 'x', total_rows is one more than total index\n",
    "            seqblock_parsed.interp_mutations = new_format\n",
    "        elif (i == total_rows - 2): # 2nd last is subjective consensus\n",
    "            seqblock_parsed.interp_consensus = new_format\n",
    "        elif (i == total_rows - 3): \n",
    "            seqblock_parsed.interp_changes = new_format\n",
    "        elif (i > 0 and i < number_of_reads + 1):\n",
    "            seqblock_parsed.reads.append(new_format)\n",
    "        else:\n",
    "            seqblock_parsed.qscores.append(new_format)\n",
    "    return seqblock_parsed     \n",
    "    \n",
    "class CleanSeqBlock():\n",
    "    def __init__(self):\n",
    "        self.reads_count = 0;\n",
    "        self.seq_len = 0;\n",
    "        \n",
    "        self.target = 0;\n",
    "        \n",
    "        self.reads = [];\n",
    "        self.qscores = [];\n",
    "        \n",
    "        self.interp_changes = 0; # indel in between y's and z shows a KNOWN base error\n",
    "        self.interp_consensus = 0;\n",
    "        self.interp_mutations = 0;\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "c7284816",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the mat file in\n",
    "file_name = 'consensus1.mat'\n",
    "variable_of_interest = 'align3'\n",
    "h5py_object, data = load_matlab_file(file_name, variable_name=variable_of_interest)\n",
    "\n",
    "DNA_seqs = DNA_SeqBlocks(h5py_object=h5py_object, data=data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "891f0e33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST FOR SEQBLOCK_PARSER\n",
    "test_1 = DNA_seqs.get_seqblock(1) # Let's look at the first DNA sequence block\n",
    "test_parse = seqblock_parser(test_1) # Let's test the parsing function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "03d0424a",
   "metadata": {},
   "outputs": [],
   "source": [
    "test_quality = test_parse.qscores[0] #Let's look at the first quality strand\n",
    "test_read = test_parse.reads[2] # Let's look at the first read strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "d3a9341f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss Checker for 3 reads:\n",
      "Success in 0th read!\n",
      "Success in 1th read!\n",
      "Success in 2th read!\n"
     ]
    }
   ],
   "source": [
    "#Read strand Testing for info loss\n",
    "\n",
    "\n",
    "print(f\"Loss Checker for {test_parse.reads_count} reads:\")\n",
    "for j in range(test_parse.reads_count):\n",
    "    Ccount = 0\n",
    "    Tcount = 0\n",
    "    Acount = 0\n",
    "    Gcount = 0\n",
    "    indelcount = 0\n",
    "    for i in test_parse.reads[j]: #Input change here\n",
    "        if i == b'C':\n",
    "            Ccount+=1;\n",
    "        if i == b'A':\n",
    "            Acount+=1;\n",
    "        if i == b'T':\n",
    "            Tcount +=1;\n",
    "        if i == b'G':\n",
    "            Gcount +=1;\n",
    "        if i == b'-':\n",
    "            indelcount +=1;\n",
    "    # print(f\" C: {Ccount} \\n G: {Gcount} \\n A: {Acount} \\n T: {Tcount} \\n indels: {indelcount}\")\n",
    "    # print(test_parse.seq_len)\n",
    "    assert Ccount+Tcount+Gcount+Acount+indelcount == test_parse.seq_len, f\"Info lost in {j}th index read \"\n",
    "    print(f\"Success in {j}th read!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "58dfda6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.15848931924611134\n",
      "1.584893192461114e-06\n",
      "1258.9254117941675\n",
      "0.07943282347242814\n",
      "7.943282347242822e-05\n",
      "63.09573444801933\n",
      "1.9952623149688787e-06\n",
      "5.011872336272725e-06\n",
      "2.5118864315095822e-05\n",
      "7.943282347242822e-07\n",
      "0.012589254117941675\n",
      "7.943282347242822e-07\n",
      "0.025118864315095794\n",
      "0.000630957344480193\n",
      "0.5011872336272722\n"
     ]
    }
   ],
   "source": [
    "# Quality Strand Testing\n",
    "#//TODO: Can I convert each char into a quality score value?\n",
    "\n",
    "def error_calculator(list_qscores):\n",
    "    for i in range(len(list_qscores)): #often more than one\n",
    "        for n in list_qscores[i]:\n",
    "            if (n == b'~'): continue\n",
    "            q = ord(n)-64\n",
    "            P = pow(10, -q/10)\n",
    "            print(P)\n",
    "            \n",
    "        \n",
    "error_calculator(test_parse.qscores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e28462af",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.9.6 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
